fdbserver/workloads/WriteTagThrottling.actor.cpp

```diff
@@ -0,0 +1,262 @@
+/*
+ * WriteThrottling.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/TesterInterface.actor.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/TagThrottle.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+// workload description:
+// This workload aims to test whether we can throttling some bad clients that doing penetrating write on write hot-spot
+// range. There are several good clientActor just randomly do read and write ops in transaction. Also, some bad
+// clientActor has high probability to read and write a particular hot-spot. If the tag-based throttling works right on
+// write-heavy tags, it will only limit the bad clientActor without influence other normal actors. This workload also
+// output TPS and latency of read/set/clear operations to do eyeball check. We want this new feature would not cause
+// more load to the cluster (Maybe some visualization tools is needed to show the trend of metrics)
+struct WriteTagThrottlingWorkload : KVWorkload {
+	// Performance metrics
+	int goodActorTxNum = 0, goodActorRetries = 0, goodActorTooOldRetries = 0, goodActorCommitFailedRetries = 0;
+	int badActorTxNum = 0, badActorRetries = 0, badActorTooOldRetries = 0, badActorCommitFailedRetries = 0;
+	double badActorTotalLatency = 0.0, goodActorTotalLatency = 0.0;
+	std::unique_ptr<ContinuousSample<double>> badActorReadLatency, goodActorReadLatency;
+	std::unique_ptr<ContinuousSample<double>> badActorCommitLatency, goodActorCommitLatency;
```

Comment:
There is no need for `std::unique_ptr` here, you can just use a `ContinuousSample<double>` field.